File: frontend/src/servicios/inventario.ts
/**
 * Cliente de inventario (T049, US2 · US7). Registrar entradas por compra, consultar existencias
 * (incluido el saldo negativo) y el listado de capital inmovilizado.
 */

import { clienteHttp } from "./clienteHttp";

export interface Lote {
  id_lote: number;
  id_producto: number;
  id_sucursal: number;
  cantidad_restante: number;
  costo_unitario: string;
  moneda: string;
  fecha_caducidad: string | null;
  instante_entrada: string;
}

export interface Existencia {
  id_sucursal: number;
  id_producto: number;
  cantidad: number;
  es_granel: boolean;
}

/** US14: un lote del desglose de existencia de un producto. `costo_unitario` sólo llega si se
 *  pidió `incluir_costo` — la pantalla de Venta nunca lo pide. */
export interface LoteExistencia {
  id_lote: number;
  cantidad: number;
  fecha_caducidad: string | null;
  instante_entrada: string;
  costo_unitario?: string | null;
}

export interface CapitalInmovilizado {
  id_lote: number;
  id_producto: number;
  id_sucursal: number;
  cantidad_restante: number;
  dias_sin_salida: number;
  dias_umbral_aplicado: number;
  umbral_heredado_del_global: boolean;
  valor_calculable: boolean;
  valor_inmovilizado: string | null;
}

export function registrarEntrada(datos: {
  id_sucursal: number;
  id_producto: number;
  cantidad: number;
  costo_unitario: string;
  fecha_caducidad?: string | null;
}): Promise<Lote> {
  return clienteHttp.post<Lote>("/entradas-inventario", datos);
}

export function listarExistencias(idSucursal: number, idProducto?: number): Promise<Existencia[]> {
  const q = idProducto !== undefined ? `&id_producto=${idProducto}` : "";
  return clienteHttp.get<Existencia[]>(`/existencias?id_sucursal=${idSucursal}${q}`);
}

/** US14: desglose por lote del saldo de un producto en una sucursal, en orden FEFO.
 *  `incluirCosto` sólo lo usan superficies con derecho a ver margen (traspasos), nunca Venta. */
export function existenciasPorLote(
  idSucursal: number,
  idProducto: number,
  incluirCosto = false,
): Promise<LoteExistencia[]> {
  const costo = incluirCosto ? "&incluir_costo=true" : "";
  return clienteHttp.get<LoteExistencia[]>(
    `/existencias/lotes?id_sucursal=${idSucursal}&id_producto=${idProducto}${costo}`,
  );
}

export function listarCapitalInmovilizado(idSucursal: number): Promise<CapitalInmovilizado[]> {
  return clienteHttp.get<CapitalInmovilizado[]>(
    `/capital-inmovilizado?id_sucursal=${idSucursal}`,
  );
}

File: backend/rasero/api/inventario.py
"""Endpoints de inventario:

- `POST /entradas-inventario` (T046, US2) — registrar una compra a proveedor.
- `GET /existencias` (T047, US2) — saldo por producto, expone el negativo tal cual.
- `GET /capital-inmovilizado` (T082, US7) — lotes parados más allá del umbral de su categoría.
"""

from datetime import date
from decimal import Decimal

from fastapi import APIRouter, Depends, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from rasero.persistencia.sesion import obtener_sesion
from rasero.servicios import inventario as servicio_inventario

router = APIRouter(tags=["inventario"])


class EntradaInventarioNueva(BaseModel):
    id_sucursal: int
    id_producto: int
    cantidad: int = Field(ge=1)
    costo_unitario: Decimal
    fecha_caducidad: date | None = None


@router.post("/entradas-inventario", status_code=status.HTTP_201_CREATED)
def registrar_entrada(
    cuerpo: EntradaInventarioNueva, sesion: Session = Depends(obtener_sesion)
) -> dict:
    lote = servicio_inventario.registrar_entrada(
        sesion,
        id_sucursal=cuerpo.id_sucursal,
        id_producto=cuerpo.id_producto,
        cantidad=cuerpo.cantidad,
        costo_unitario=cuerpo.costo_unitario,
        fecha_caducidad=cuerpo.fecha_caducidad,
    )
    restante = servicio_inventario.cantidad_restante_lote(sesion, lote.id_lote)
    return {
        "id_lote": lote.id_lote,
        "id_producto": lote.id_producto,
        "id_sucursal": lote.id_sucursal,
        "cantidad_restante": int(restante),
        "costo_unitario": f"{Decimal(lote.costo_unitario):.4f}",
        "moneda": lote.moneda,
        "fecha_caducidad": lote.fecha_caducidad.isoformat() if lote.fecha_caducidad else None,
        "instante_entrada": lote.instante_entrada,
    }


@router.get("/existencias")
def listar_existencias(
    id_sucursal: int,
    id_producto: int | None = None,
    sesion: Session = Depends(obtener_sesion),
) -> list[dict]:
    return servicio_inventario.listar_existencias(
        sesion, id_sucursal=id_sucursal, id_producto=id_producto
    )


@router.get("/existencias/lotes")
def existencias_por_lote(
    id_sucursal: int,
    id_producto: int,
    incluir_costo: bool = False,
    sesion: Session = Depends(obtener_sesion),
) -> list[dict]:
    """Desglose por lote del saldo de un producto en una sucursal, en orden FEFO (US14).
    `incluir_costo` sólo lo usan las superficies con derecho a ver margen (traspasos de un
    encargado), nunca la pantalla de Venta.
    """
    return servicio_inventario.existencia_por_lote(
        sesion,
        id_sucursal=id_sucursal,
        id_producto=id_producto,
        incluir_costo=incluir_costo,
    )


@router.get("/capital-inmovilizado")
def capital_inmovilizado(
    id_sucursal: int | None = None, sesion: Session = Depends(obtener_sesion)
) -> list[dict]:
    return servicio_inventario.capital_inmovilizado(sesion, id_sucursal=id_sucursal)
